cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -394,15 +394,30 @@ public synchronized void stopAdapter(Id.Namespace namespace, String adapterName)
    * @throws IOException if there was some error starting worker adapter
    */
   public synchronized void startAdapter(Id.Namespace namespace, String adapterName)
-    throws NotFoundException, InvalidAdapterOperationException, SchedulerException, IOException {
+    throws NotFoundException, InvalidAdapterOperationException, SchedulerException, IOException, ExecutionException,
+    InterruptedException {
+    AdapterDefinition adapterSpec = getAdapter(namespace, adapterName);
+    ProgramType programType = adapterSpec.getProgram().getType();
+
     AdapterStatus adapterStatus = getAdapterStatus(namespace, adapterName);
     if (AdapterStatus.STARTED.equals(adapterStatus)) {
-      throw new InvalidAdapterOperationException("Adapter is already started.");
+      // For Adapter with type of Worker, need to check whether the program is running since Workflow should
+      // already added to scheduler
+      if (programType == ProgramType.WORKER) {
+        Id.Program program = getProgramId(namespace, adapterName);
+        ProgramRuntimeService.RuntimeInfo runtimeInfo = lifecycleService.findRuntimeInfo(program, programType);
+        // check if it is not running
+        if (runtimeInfo == null) {
+          // not running, lets update the RunRecords
+          stopWorkerAdapter(namespace, adapterSpec);
```

Comment:
this will not do anything. It will call lifecycleService.stopProgram(), which will fail to find the runtime info and then do nothing. If I understand correctly, this is the case where the worker crashed, thus its runtime info does not exist any longer, but its run record has not been updated. You need to correct its run record to FAILED, similar to what is done in the program lifecycle handler.